// @guildedts/builders/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Colors = exports.resolveColor = void 0;
/**
 * Resolve a color from a string, number, RGB array or color preset.
 * Inspired by Guilded.JS' {@link https://github.com/guildedjs/guilded.js/blob/main/packages/webhook-client/lib/util.ts#L8 color resolver}.
 * @param color The color to resolve.
 * @returns The resolved color.
 * @example
 * // Resolve the color with a hex color.
 * resolveColor(0xFFFFFF);
 * resolveColor('#FFFFFF');
 * // Resolve the color with a preset color.
 * resolveColor('WHITE');
 * // Resolve the color with RGB values.
 * resolveColor([255, 255, 255]);
 */
function resolveColor(color) {
    if (color === 'RANDOM')
        color = Math.floor(Math.random() * (0xffffff + 1));
    else if (typeof color === `string`)
        color = Colors[color] ?? parseInt(color.replace('#', ''), 16);
    else if (Array.isArray(color))
        color = (color[0] << 16) + (color[1] << 8) + color[2];
    return color;
}
exports.resolveColor = resolveColor;
/** A set of color presets. */
var Colors;
(function (Colors) {
    Colors[Colors["DEFAULT"] = 0] = "DEFAULT";
    Colors[Colors["WHITE"] = 16777215] = "WHITE";
    Colors[Colors["BLACK"] = 0] = "BLACK";
    Colors[Colors["GILDED"] = 16106496] = "GILDED";
    Colors[Colors["GREY"] = 3552829] = "GREY";
    Colors[Colors["RED"] = 14692657] = "RED";
    Colors[Colors["PINK"] = 12723548] = "PINK";
    Colors[Colors["GRAPE"] = 10237621] = "GRAPE";
    Colors[Colors["VIOLET"] = 6767065] = "VIOLET";
    Colors[Colors["INDIGO"] = 3890139] = "INDIGO";
    Colors[Colors["BLUE"] = 1667522] = "BLUE";
    Colors[Colors["CYAN"] = 820633] = "CYAN";
    Colors[Colors["TEAL"] = 627304] = "TEAL";
    Colors[Colors["GREEN"] = 3120708] = "GREEN";
    Colors[Colors["LIME"] = 6727695] = "LIME";
    Colors[Colors["YELLOW"] = 16565273] = "YELLOW";
    Colors[Colors["ORANGE"] = 15227148] = "ORANGE";
})(Colors = exports.Colors || (exports.Colors = {}));
